### PR TITLE
Enable nullable references for generated C# code

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR adds
```csharp
#nullable enable
```

to the C# files generated by slice2cs.

### CODE REVIEW

File: cpp/src/slice2cs/CsUtil.cpp [hunk 1]
```diff
@@ -318,8 +318,9 @@ Slice::CsGenerator::typeToString(const TypePtr& type, const string& package, boo
         return "void";
     }
 
-    if (optional && isValueType(type))
+    if (optional && !isProxyType(type))
     {
+        // Proxy types are mapped the same way for optional and non-optional types.
```
Comment (bernardnormier) [inline cpp/src/slice2cs/CsUtil.cpp:323]: And it can't be an optional class types since they are now disallowed.

File: cpp/src/slice2cs/Gen.cpp [hunk 2]
```diff
@@ -1820,7 +1820,8 @@ Slice::Gen::Gen(const string& base, const vector<string>& includePaths, const st
 
     printGeneratedHeader(_out, fileBase + ".ice");
 
-    _out << nl << "using _System = global::System;";
+    _out << nl << "#nullable enable";
+    _out << sp << nl << "using _System = global::System;";
```
Comment (pepone) [inline cpp/src/slice2cs/Gen.cpp:1824]: Seems we don't use `_System` anywhere, is this required?

File: cpp/src/slice2cs/Gen.cpp [hunk 3]
```diff
@@ -3950,7 +3955,7 @@ Slice::Gen::DispatcherVisitor::visitInterfaceDefStart(const InterfaceDefPtr& p)
 
         writeTieOperations(p);
 
-        _out << sp << nl << "private " << name << opIntfName << "_ _ice_delegate;";
+        _out << sp << nl << "private " << name << opIntfName << "_? _ice_delegate;";
```
Comment (bernardnormier) [inline cpp/src/slice2cs/Gen.cpp:3958]: Kept the Tie code as-is. I am in favor of removing Ties entirely in a follow-up PR. What do you think?

**Review by externl — APPROVED** [review]

File: cpp/src/slice2cs/Gen.cpp [hunk 4]
```diff
@@ -901,7 +901,7 @@ Slice::CsVisitor::getDispatchParams(
     }
 
     string currentParamName = getEscapedParamName(op, "current");
-    params.push_back(getUnqualified("Ice.Current", ns) + " " + currentParamName + " = null");
+    params.push_back(getUnqualified("Ice.Current?", ns) + " " + currentParamName + " = null");
```
Comment (bernardnormier) [inline cpp/src/slice2cs/Gen.cpp:904]: Kept all this code "as is". 

In a follow-up PR, I want to make this Current parameter non-nullable.

**Review by pepone — APPROVED** [review]

File: cpp/src/slice2cs/Gen.cpp (hunk 2 above)
Comment (bernardnormier) [reply cpp/src/slice2cs/Gen.cpp:1824]: Yes, we should remove it. It's not used anywhere.

**Pull request merged**